lnrpc/invoicesrpc/addinvoice.go

```diff
@@ -678,169 +627,213 @@ type SelectHopHintsCfg struct {
-			totalHintBandwidth,
+			nHintsLeft, currentBandwidth, targetBandwidth,
 		)
 		if enoughHopHints {
-			log.Debugf("Second pass of hop selection has " +
-				"sufficient hints")
-
 			return hopHints
 		}
 
-		channel := openChannels[i]
+		hopHint, remoteBalance, include := shouldIncludeChannel(
+			channel, alreadyIncluded, cfg,
+		)
 
-		// Skip the channel if we already selected it.
-		if _, ok := hopHintChans[channel.FundingOutpoint]; ok {
-			continue
+		if include {
+			// Now that we now this channel use usable, add it as a hop
+			// hint and the indexes we'll use later.
+			hopHints = append(hopHints, []zpay32.HopHint{hopHint})
+			currentBandwidth += remoteBalance
+			nHintsLeft--
 		}
+	}
 
-		// If the channel can't be a hop hint, then we'll skip it.
-		// Otherwise, we'll use the policy information to populate the
-		// hop hint.
-		remotePolicy, canBeHopHint := chanCanBeHopHint(channel, cfg)
-		if !canBeHopHint || remotePolicy == nil {
-			continue
-		}
+	// We do not want to leak information about how our remote balance is
+	// distributed in our private channels. We shuffle the selected ones
+	// here so they do not appear in order in the invoice.
+	mathRand.Shuffle(
+		len(hopHints), func(i, j int) {
+			hopHints[i], hopHints[j] = hopHints[j], hopHints[i]
+		},
+	)
+	return hopHints
+}
 
-		// Lookup and see if there's an alias SCID that exists.
-		chanID := lnwire.NewChanIDFromOutPoint(
-			&channel.FundingOutpoint,
-		)
-		alias, _ := cfg.GetAlias(chanID)
-
-		// If this is a channel where the option-scid-alias feature bit
-		// was negotiated and the alias is not yet assigned, we cannot
-		// issue an invoice. Doing so might expose the confirmed SCID
-		// of a private channel.
-		if channel.ScidAliasFeature {
-			var defaultScid lnwire.ShortChannelID
-			if alias == defaultScid {
-				continue
-			}
-		}
+// PopulateHopHints will select up to cfg.MaxHophints from the current open
+// channels. The set of hop hints will be returned as a slice of functional
+// options that'll append the route hint to the set of all route hints.
+//
+// TODO(roasbeef): do proper sub-set sum max hints usually << numChans.
+func PopulateHopHints(amtMSat lnwire.MilliSatoshi, cfg *SelectHopHintsCfg,
+	forcedHints [][]zpay32.HopHint) ([][]zpay32.HopHint, error) {
 
-		// Include the route hint in our set of options that will be
-		// used when creating the invoice.
-		addHopHint(&hopHints, channel, remotePolicy, alias)
+	hopHints := forcedHints
 
-		// As we've just added a new hop hint, we'll accumulate it's
-		// available balance now to update our tally.
-		//
-		// TODO(roasbeef): have a cut off based on min bandwidth?
-		totalHintBandwidth += channel.RemoteBalance
+	// If we already have enough hints we don't need to add any more.
+	nHintsLeft := cfg.MaxHopHints - len(hopHints)
+	if nHintsLeft <= 0 {
+		return hopHints, nil
 	}
 
-	return hopHints
+	alreadyIncluded := make(map[uint64]struct{})
```

Comment:
How about instead using `map[uint64]bool` for `alreadyIncluded` instead? 

That makes the logic later easier to read cause you can do 
```
if alreadyIncluded[xxx] {
```
instead of 
```
_, ok := alreadyIncluded[xxx]
if ok {
```

Reply:
ah! TIL - awesome, good to know!